File: Largest-Rectangle-in-Histogram.py
class Solution:
    def largestRectangleArea(self, heights):
        res = 0
        n = len(heights)
        lfs = [0] * n
        rfs = [n - 1] * n
        stack = []

        for i in range(n):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            if stack:
                lfs[i] = stack[-1] + 1
            else:
                lfs[i] = 0
            stack.append(i)

        stack = []  

        for i in range(n - 1, -1, -1):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            if stack:
                rfs[i] = stack[-1] - 1
            else:
                rfs[i] = n - 1
            stack.append(i)

        for i in range(n):
            res = max(res, heights[i] * (rfs[i] - lfs[i] + 1))

        return res

File: Largest-Rectangle-in-Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int res = 0;
        int n = heights.size();
        vector<int> lfs(n);
        vector<int> rfs(n, n - 1);
        stack<int> s;

        for (int i = 0; i < n; i++) {
            while (!s.empty() && heights[s.top()] >= heights[i]) {
                s.pop();
            }
            if (!s.empty()) {
                lfs[i] = s.top() + 1;
            } else {
                lfs[i] = 0;
            }
            s.push(i);
        }

        while (!s.empty()) {
            s.pop();
        }

        for (int i = n - 1; i >= 0; i--) {
            while (!s.empty() && heights[s.top()] >= heights[i]) {
                s.pop();
            }
            if (!s.empty()) {
                rfs[i] = s.top() - 1;
            } else {
                rfs[i] = n - 1;
            }
            s.push(i);
        }

        for (int i = 0; i < n; i++) {
            res = max(res, heights[i] * (rfs[i] - lfs[i] + 1));
        }

        return res;
    }
};
